craft_application/services/lifecycle.py

```diff
@@ -290,10 +294,7 @@ def _update_project_variables(self) -> None:
         """Replace project fields with values set using craftctl."""
         update_vars: dict[str, str] = {}
         for var in self._app.project_variables:
-            value = self.project_info.get_project_var(var)
-            if not value:
-                raise errors.PartsLifecycleError(f"Project field '{var}' was not set.")
-            update_vars[var] = value
+            update_vars[var] = self.project_info.get_project_var(var)
```

Comment:
So this one allowed to be empty?

Reply:
This verification was moved to a different place to handle a list of application-defined mandatory fields. We can't do it here because this list is not available in the project model.

Reply:
What if we made the set of adoptable fields a property of the project model rather than the application? Then we could have separate adoptable lists for core22 projects and core24 projects by making the project we load be a tagged union.


If we created a pydantic field type like:

```
Adoptable = pydantic.Field(allow_none=True, adoptable=True)
```

Then we could have two models:

```
class Core22Project(SnapProject):
    some_field: Annotated[str | None, Adoptable]
    ...
```

and:

```
class Core24Project(SnapProject):
    some_field: Annotated[str, Adoptable]
```

Then we could inspect the class's fields to find adoptable ones and ensure the adopted variable is a valid instance of the type. In my example above, an unset field would return None, so it would be valid for core22 but invalid for core24.

Reply:
In this case, for symmetry, project variables can be also specified in the project, which makes sense given the name. But we may want to land this to unblock Sheng and then refactor in a different PR.